extensions/amp-analytics/0.1/visibility-helper.js

```diff
@@ -0,0 +1,397 @@
+   *  3. The parent's visibility.
+   * @return {number}
+   */
+  getVisibility() {
+    const ownVisibility = this.ownVisibility_ * (this.blocked_ ? 0 : 1);
+    if (!this.parent_) {
+      return ownVisibility;
+    }
+    if (this.factorParent_) {
+      return ownVisibility * this.parent_.getVisibility();
+    }
+    return ownVisibility * (this.parent_.getVisibility() > 0 ? 1 : 0);
+  }
+
+  /**
+   * Runs the calculation cycle.
+   */
+  update() {
+    const visibility = this.getVisibility();
+    this.update_(visibility);
+    if (this.children_) {
+      for (let i = 0; i < this.children_.length; i++) {
+        this.children_[i].update();
+      }
+    }
+  }
+
+  /**
+   * Returns the calculated state of visibility.
+   * @param {time} startTime
+   * @return {!Object<string, string|number>}
+   */
+  getState(startTime) {
+    return {
+      // Observed times, relative to the `startTime`.
+      firstSeenTime: timeBase(this.firstSeenTime_, startTime),
+      lastSeenTime: timeBase(this.lastSeenTime_, startTime),
+      lastVisibleTime: timeBase(this.lastVisibleTime_, startTime),
+      fistVisibleTime: timeBase(this.fistVisibleTime_, startTime),
+
+      // Durations.
+      maxContinuousVisibleTime: this.maxContinuousVisibleTime_,
+      totalVisibleTime: this.totalVisibleTime_,
+
+      // Visibility percents.
+      loadTimeVisibility: this.loadTimeVisibility_ * 100 || 0,
+      minVisiblePercentage: this.minVisiblePercentage_ * 100,
+      maxVisiblePercentage: this.maxVisiblePercentage_ * 100,
+    };
+  }
+
+  /**
+   * @param {number} visibility
+   * @private
+   */
+  update_(visibility) {
+    if (!this.eventResolver_) {
+      return;
+    }
+    // Update state and check if all conditions are satisfied
+    const conditionsMet = this.updateCounters_(visibility);
+    if (conditionsMet) {
+      if (this.scheduledRunId_) {
+        clearTimeout(this.scheduledRunId_);
+        this.scheduledRunId_ = null;
+      }
+      this.eventResolver_();
+    } else if (this.matchesVisibility_ && !this.scheduledRunId_) {
+      // There is unmet duration condition, schedule a check
+      const timeToWait = this.computeTimeToWait_();
+      if (timeToWait > 0) {
+        this.scheduledRunId_ = setTimeout(() => {
+          this.scheduledRunId_ = null;
+          this.update();
+        }, timeToWait);
+      }
+    } else if (!this.matchesVisibility_ && this.scheduledRunId_) {
+      clearTimeout(this.scheduledRunId_);
+      this.scheduledRunId_ = null;
+    }
+  }
+
+  /**
+   * @param {number} visibility
+   * @return {boolean} true
+   * @private
+   */
+  updateCounters_(visibility) {
+    const now = Date.now();
+
+    if (visibility > 0) {
+      this.firstSeenTime_ = this.firstSeenTime_ || now;
+      this.lastSeenTime_ = now;
+      // Consider it as load time visibility if this happens within 300ms of
+      // page load.
+      if (!this.loadTimeVisibility_ && (now - this.createdTime_) < 300) {
```

Comment:
No, not certain. But this is ok. This is a very rough metric and the legacy code supports it the same way.